### PR TITLE
Features test sahf

### PR DESCRIPTION
**Pull request by watchdogoblivion**
Added full coverage for a majority of files. Rewrote broken tests and implemented tests that were not written from additional pulls from dev.

### CODE REVIEW

**Comment by carolynrehm** [comment]
@IkenoXamos

File: src/app/components/content-creator-page/content-creator-page.component.spec.ts [hunk 1]
```diff
@@ -121,104 +120,186 @@ describe('ContentCreatorPageComponent', () => {
   });
   // =====End Element creation finds=====
 
-  /**
```
Comment (IkenoXamos) [inline src/app/components/content-creator-page/content-creator-page.component.spec.ts:124]: How come all relevant comments have been removed? If there have been changes which make the comment incorrect, it should be updated. There is no need to remove entirely. This whole file has a lot more tests, which are good, but all documentation is now gone. This needs documentation.

File: src/app/components/content-finder-page/content-finder-page.component.spec.ts [hunk 2]
```diff
@@ -42,61 +42,63 @@ describe('ContentFinderPageComponent', () => {
     component.selectedSubjects = ['Java', 'CSS'];
   });
 
-  /**
```
Comment (IkenoXamos) [inline src/app/components/content-finder-page/content-finder-page.component.spec.ts:45]: Same issue, all comments have been removed. Some amount of comments/documentation is important.

File: src/app/components/curriculum-creator-page/curriculum-creator-page.component.spec.ts [hunk 3]
```diff
@@ -51,19 +59,556 @@ describe('CurriculumCreatorPageComponent', () => {
...
+    curriculumModule2 = new CurriculumModule(2,2,2,null);
+    curriculumModules = [];
+    mockEvent = {currentIndex: 0, previousIndex: 0}
     fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
```
Comment (IkenoXamos) [inline src/app/components/curriculum-creator-page/curriculum-creator-page.component.spec.ts:86]: Lots of tests, which is nice to have. However, there is once again zero comments explaining what is going on. This is important.

File: src/app/components/curriculum-creator-page/curriculum-creator-page.component.spec.ts [hunk 4]
```diff
@@ -51,19 +59,556 @@ describe('CurriculumCreatorPageComponent', () => {
...
+      observer.complete();
+    });
+    spyOn(component.cfs,'createCurriculum').and.returnValue(observable);
+    component.submit();
+    expect(toastrService.previousToastMessage).toBe('Failed to communicate with the server');
+  });
+
+});
+
+describe('DeleteCurriculumDialog', () => {
```
Comment (IkenoXamos) [inline src/app/components/curriculum-creator-page/curriculum-creator-page.component.spec.ts:427]: There should perhaps be comments explaining the goal of each group of tests at the corresponding describe. Further explanation can be provided at the beforeEach, as well as each test as necessary. Preferably at least the first test should have a bit of an explanation, and if the following tests are similar enough, they don't need one. But at least one comment for the first test explaining what is happening.

File: src/app/components/module-creator-page/module-creator-page.component.spec.ts [hunk 5]
```diff
@@ -1,85 +1,142 @@
...
   });
 
-  // Testing to make sure that we have good inputs
   it('Should return true if Module field is valid', () => {
     expect(component.validInput()).toBeTruthy();
   });
 
-  // Testing to make sure the resetVariables function works
-  it('Should reset the fields with the resetVariables function', () => {
+  it('Should reset the fields with the resetVariables function, subject', () => {
```
Comment (IkenoXamos) [inline src/app/components/module-creator-page/module-creator-page.component.spec.ts:82]: Comments/Documentation.

File: src/app/components/module-index-page/module-index-page.component.spec.ts [hunk 6]
```diff
@@ -1,138 +1,365 @@
...
-    let mod2: Module;
-    let mod3: Module;
+  beforeEach(() => {
+    attributes = {id: {nodeValue: "node"}}
+    dummyEvent = {target: {attributes: attributes}, srcElement: {attributes: attributes}, currentTarget: {attributes: attributes}}
+    fixture.detectChanges();
+  });
 
-    beforeEach(async(() => {
+  it('should create', () => {
```
Comment (IkenoXamos) [inline src/app/components/module-index-page/module-index-page.component.spec.ts:73]: Comments/Documentation.

File: src/app/components/reports-page/reports-page.component.spec.ts [hunk 7]
```diff
@@ -87,54 +79,140 @@ describe('ReportsPageComponent', () => {
...
   });
 
-// Test that formats are defined as expected
   it('should have defualt formats defined', () => {
     expect(component.formats).toEqual(["Code", "Document", "Powerpoint", "All"]);
   });
 
-// Test that metrics functions work
-  it('should have metrics defined after ngOnInit()', () => {
+  it('should test ngOnInit() updateMetrics', () => {
```
Comment (IkenoXamos) [inline src/app/components/reports-page/reports-page.component.spec.ts:94]: Comments/Documentation.

File: src/app/components/reports-page/reports-page.component.ts [hunk 8]
```diff
@@ -1,8 +1,5 @@
 import { Component, OnInit } from '@angular/core';
-import { HttpClient } from '@angular/common/http';
```
Comment (IkenoXamos) [inline src/app/components/reports-page/reports-page.component.ts:2]: Be careful about removing imports in PRs that are not about cleaning/refactoring. This is okay for now, but in the future, even if they can be cleaned/removed, leave them be for a PR that is dedicated to cleaning/refactoring this type of thing.

File: src/app/components/reports-time-graph/reports-time-graph.component.spec.ts [hunk 9]
```diff
@@ -126,4 +116,45 @@ describe('ReportsTimeGraphComponent', () => {
     expect(component.selectedView).toEqual("Past Year");
   });
 
+  it('should test setGraphResults datnum > startime', () => {
```
Comment (IkenoXamos) [inline src/app/components/reports-time-graph/reports-time-graph.component.spec.ts:119]: Comments/Documentation.

File: src/app/components/resolve-request-page/resolve-request-page.component.spec.ts [hunk 10]
```diff
@@ -24,20 +45,198 @@ describe('ResolveRequestPageComponent', () => {
...
-    fixture = TestBed.createComponent(ResolveRequestPageComponent);
-    component = fixture.componentInstance;
     fixture.detectChanges();
   });
 
-  it('should create', () => {
+  it('should create component', () => {
     expect(component).toBeTruthy();
   });
+
```
Comment (IkenoXamos) [inline src/app/components/resolve-request-page/resolve-request-page.component.spec.ts:80]: Comments/Documentation.

File: src/app/components/submit-request-page/submit-request-page.component.spec.ts [hunk 11]
```diff
@@ -1,40 +1,139 @@
...
   beforeEach(() => {
-    fixture = TestBed.createComponent(SubmitRequestPageComponent);
-    component = fixture.componentInstance;
     fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
```
Comment (IkenoXamos) [inline src/app/components/submit-request-page/submit-request-page.component.spec.ts:56]: Comments/Documentation.

File: src/app/pipes/curriculum-module-filter.pipe.spec.ts [hunk 12]
```diff
@@ -1,8 +1,50 @@
...
+    modules = [];
+    curriculum = new Curriculum(1, "Java", []);
+    module = new Module(1, "Java", 12345,null,null, null, null);
+    curriculumModule = new CurriculumModule(1, 1, 1, module);
+  });
+
   it('create an instance', () => {
-    const pipe = new CurriculumModuleFilterPipe();
-    expect(pipe).toBeTruthy();
+    expect(curriculumModuleFilterPipe).toBeTruthy();
```
Comment (IkenoXamos) [inline src/app/pipes/curriculum-module-filter.pipe.spec.ts:29]: Comments/Documentation.

File: src/app/services/content-fetcher.service.spec.ts [hunk 13]
```diff
@@ -26,133 +25,98 @@ describe('ContentFetcherService', () => {
...
-      (error: any) => { }
-    );
+    service.createNewContent(new Content(null, null, null, null, null, null)).subscribe();
     const req = httpTestingController.expectOne(baseURL + '/content');
     expect(req.request.method).toEqual("POST");
-    req.flush(response);
     tick();
+  }));
 
+   // Testing addLinkToContent(content) function
```
Comment (IkenoXamos) [inline src/app/services/content-fetcher.service.spec.ts:59]: At least there is a comment here, but perhaps a little bit more detail would be helpful. The next group of associates who work on this might want to also know a bit about the methodology that you are using to test. This also applies to all testing related comments.

File: src/app/services/curriculum-fetcher.service.spec.ts [hunk 14]
```diff
@@ -1,27 +1,78 @@
...
-
+    curriculumFetcherService = TestBed.get(CurriculumFetcherService);
+    curriculum = new Curriculum(1,"Java",[]);
+    curriculumModule = new CurriculumModule(1,1,1,null);
   });
 
   it('should be created', () => {
-    const service: CurriculumFetcherService = TestBed.get(CurriculumFetcherService);
-    expect(service).toBeTruthy();
+    expect(curriculumFetcherService).toBeTruthy();
```
Comment (IkenoXamos) [inline src/app/services/curriculum-fetcher.service.spec.ts:31]: Comments/Documentation.

File: src/app/services/curriculum-store.service.spec.ts [hunk 15]
```diff
@@ -1,23 +1,91 @@
...
+    toastrService = TestBed.get(ToastrService);
+    curriculumModule = new CurriculumModule(1,1,1,null);
+    curriculum = new Curriculum(1,"Java", []);
+    curriculums = [];
+  });
 
   it('should be created', () => {
-    const service: CurriculumStoreService = TestBed.get(CurriculumStoreService);
-    expect(service).toBeTruthy();
+    expect(curriculumStoreService).toBeTruthy();
```
Comment (IkenoXamos) [inline src/app/services/curriculum-store.service.spec.ts:39]: Comments/Documentation.

File: src/app/services/module-fetcher.service.spec.ts [hunk 16]
```diff
@@ -1,194 +1,126 @@
...
-  /** 
-    * First stipulate the fakeAsync() 
-    * function to generate an empty resposne that one will manually 
-    * populate. Using service.getAllModules().subscribe() will give a 
-    * mocked response. Lastly do a test as necessary to test if the method
-    * is indeed a "GET" method, and use req.flush(response) and tick() to 
-    * finalize mock.
-    */
-  it('should return an Observable<Module[]> Get all', fakeAsync(() => {
+  it('should test getAllModules', fakeAsync(() => {
```
Comment (IkenoXamos) [inline src/app/services/module-fetcher.service.spec.ts:36]: Comments/Documentation.

File: src/app/services/module-fetcher.service.spec.ts [hunk 17]
```diff
@@ -1,194 +1,126 @@
...
+    const req = httpTestingController.expectOne(url);
+    expect(req.request.method).toEqual("DELETE");
+    req.flush(null);
+    tick();
   }));
 
-  // A getAllFakeData method is needed in the project, so mocking is needed
-  // once again.
-  it('should return an Observable<Module[]> Get all fake data', fakeAsync(() => {
+  it('should test getAllFakeModules', fakeAsync(() => {
```
Comment (IkenoXamos) [inline src/app/services/module-fetcher.service.spec.ts:108]: What is fakeAsync used for? Perhaps a comment/documentation explaining such would be helpful.

File: src/app/services/module-store.service.spec.ts [hunk 18]
```diff
@@ -17,61 +16,177 @@ describe('ModuleStoreService', () => {
...
-   * First, we test the LoadModules functiong. To do so, mocking is needed.
-   * First stipulate the fakeAsync()
-   * function to generate an empty response that one will manually
-   * populate.
-   */
-  it('loadModules should be working correctly', fakeAsync(() => {
-    let response = {};
-
-    service.loadModules();
+  it('loadModules response not null', fakeAsync(() => {
```
Comment (IkenoXamos) [inline src/app/services/module-store.service.spec.ts:38]: Comments/Documentation.

File: src/app/services/pages.service.spec.ts [hunk 19]
```diff
@@ -1,18 +1,32 @@
...
 
-    expect(service).toBeTruthy();
-  
+  it('test constructor', () => {
+    pagesService = new PagesService();
+    expect(pagesService).toBeDefined();
   });
 
+  // Test only serves the purpose of full coverage since we are unable to spy on window since it is global.
+  // Also normal attmpt to test window.location.reload ends in loop.
```
Comment (IkenoXamos) [inline src/app/services/pages.service.spec.ts]: Perhaps more information can be provided about this. Future associates may need to fix such issues in a different sprint. More information about issues is always helpful.

File: src/app/services/pages.service.ts [hunk 20]
```diff
@@ -4,17 +4,17 @@ import { Injectable } from '@angular/core';
   providedIn: 'root'
 })
 export class PagesService {
-
+  window:any = window;
```
Comment (IkenoXamos) [inline src/app/services/pages.service.ts]: Is this necessary? Window is already global, so it seems unnecessary to have an instance which points to it.

File: src/app/services/reports.service.spec.ts [hunk 21]
```diff
@@ -1,52 +1,156 @@
...
   });
 
-  /**
-   * There was an attempt to test the getMetrics() method here, however we were unable
-   * to figure out how to define reportsPage.updateMetrics() and reportsTimeGraph.updateGraph()
-   * in this spec.ts file. When running a test on that method it would fail saying reportsPage and 
-   * reportsTimeGraph were undefined. We unfortunately were unable to properly test this function in 
-   * the time we had. 
-   */
+  it('getMetricsTest, test subscribe, metricsData', () => {
```
Comment (IkenoXamos) [inline src/app/services/reports.service.spec.ts:55]: Comments/Documentation.

File: src/app/services/reports.service.ts [hunk 22]
```diff
@@ -19,7 +19,7 @@ export class ReportsService {
   private readonly HEADERS = new HttpHeaders({ 'Content-Type': 'application/json' });
   
   /** Milliseconds per year */
-  private readonly MILLIS_PER_YEAR: number = 3.154e+10;
+  readonly MILLIS_PER_YEAR: number = 3.154e+10;
```
Comment (IkenoXamos) [inline src/app/services/reports.service.ts:22]: Any reason why this needs to be non-private?

File: src/app/services/request-fetcher.service.spec.ts [hunk 23]
```diff
@@ -1,18 +1,79 @@
...
+    service = TestBed.get(RequestFetcherService);
+    httpTestingController = TestBed.get(HttpTestingController);
+    baseURL = environment.cms_url;
+  });
 
   it('should be created', () => {
-    const service: RequestFetcherService = TestBed.get(RequestFetcherService);
     expect(service).toBeTruthy();
   });
+
```
Comment (IkenoXamos) [inline src/app/services/request-fetcher.service.spec.ts:29]: Comments/Documentation.

File: src/app/services/sort-search.service.spec.ts [hunk 24]
```diff
@@ -1,12 +1,311 @@
...
+      curriculumModule2 = new CurriculumModule(2,2,2,null);
+      curriculums = [];
+    }
+  );
 
   it('should be created', () => {
-    const service: SortSearchService = TestBed.get(SortSearchService);
     expect(service).toBeTruthy();
   });
+
```
Comment (IkenoXamos) [inline src/app/services/sort-search.service.spec.ts:40]: Comments/Documentation.

File: src/app/services/sort-search.service.ts [hunk 25]
```diff
@@ -27,7 +27,7 @@ export class SortSearchService {
 
             return null;
         }
-
+
```
Comment (IkenoXamos) [inline src/app/services/sort-search.service.ts]: Is this change really necessary? You're adding a tab to an empty line. This file should be removed from the PR.

File: src/app/services/submit-request.service.spec.ts [hunk 26]
```diff
@@ -1,19 +1,35 @@
...
+    service = TestBed.get(SubmitRequestService);
+    httpTestingController = TestBed.get(HttpTestingController);
+    baseURL = environment.cms_url;
+  });
 
   it('should be created', () => {
-    const service: SubmitRequestService = TestBed.get(SubmitRequestService);
     expect(service).toBeTruthy();
   });
+
```
Comment (IkenoXamos) [inline src/app/services/submit-request.service.spec.ts:27]: Comments/Documentation.

**Review by IkenoXamos — CHANGES_REQUESTED** [review]
There are a lot of tests which appear to passing. Additionally, from a glance, they appear to be decent tests. However, an almost complete lack of comments/documentation needs to be fixed. Future associates will have a difficult time managing these tests without them.

File: src/app/components/content-creator-page/content-creator-page.component.spec.ts (hunk 1 above)
Comment (watchdogoblivion) [reply src/app/components/content-creator-page/content-creator-page.component.spec.ts:124]: Hello @IkenoXamos ,  I removed the comments to adhere to best practice. As stated by various developers and by Carolyn, unit test names should be explanatory in themselves. comments should be added mainly where additional information may be necessary. The first argument in a unit test function, the string, should clearly describe what a test is meant to do. The additional comments may have been necessary when previous tests had multiple assertions but having multiple assertions in unit tests is itself also bad practice. Looking at the names I have added, I realize the string argument can be written better. However, by adhering to best practice, I still do not see the need for a lot of the comments. In addition to this, if the comments were to be kept, I learned that they should be used as document comments rather than pseudo like comments. Like below:
/**
  * Gets script version
  * @param fileName
  * @returns script version
  */
A good example of why I removed comments is the comment "// Test radio selformat" which is placed on top of an assertion in a method that has multiple assertions. FIrst I split these assertions into multiple methods then I added the description in it. So there is no point in having a comment on that. 
Most of my references are documentation and/or Carolyn. All unit tests should adhere to the FIRST principle. One of the principles states "A test method should test for a single logical outcome, implying that typically there
should be only a single logical assert.". Following that logic, and usage of the first parameter in "it", there seems to be little need for a majority of the comments that I saw. 
In the comments, the use of the words "now", "first", "next", implies an ordering to the tests. However tests should run independent of ordering. That may not be what was intended but simply how I perceived it.
As I previously stated, I do believe my test descriptions could possibly be written better, so I can look into renaming them.

Comment (IkenoXamos) [reply src/app/components/content-creator-page/content-creator-page.component.spec.ts:124]: Yes, that in general is all fine. I agree with those best practices for the most part. However, I suppose I am more of an opinion that some header comments using the /** are acceptable. I wasn't suggesting that each individual it() has its own header comment either. But, some of your tests were not quite as simple as a single assert, and used external libraries such as spyOn. These libraries are common, so not a whole lot of explanation would be needed, but a little bit is helpful. There are some test files with hundreds of lines of tests, and no extra documentation outside of the it() statements. Even if those it() statements were clarified, there are definitely a good several spots where even further documentation in the form of /** comments would be very helpful.

Comment (carolynrehm) [reply src/app/components/content-creator-page/content-creator-page.component.spec.ts:124]: This is a good point with the external libraries. There is a testing section in the wiki, maybe we can include some general documentation there to help shed some light on how additional libraries were used for future associates.

File: src/app/services/pages.service.spec.ts (hunk 19 above)
Comment (carolynrehm) [reply src/app/services/pages.service.spec.ts]: Perhaps we can document this on asana?

File: src/app/components/content-creator-page/content-creator-page.component.spec.ts (hunk 1 above)
Comment (watchdogoblivion) [reply src/app/components/content-creator-page/content-creator-page.component.spec.ts:124]: The wiki doc sounds good. Documenting when it is necessary to use fake async and tick to handle the microtasks would probably be good since there is a constant shift between promises and observables within the code. Also when Michael added the Mat* related libraries, I ended up needing to pass in custom spy objects into the provider's values in order to properly use the dependencies for testing. If there is a location that requires specific documentation, I will gladly add it.

File: src/app/components/reports-time-graph/reports-time-graph.component.ts [hunk 27]
```diff
@@ -62,7 +62,7 @@ export class ReportsTimeGraphComponent implements OnInit {
   selectedView: string = null;
 
   /** Viewport base width of the line chart */
-  w: number = Math.max(document.documentElement.clientWidth, window.innerWidth || 0);
+  w: number = Math.max(document.documentElement.clientWidth, window.innerWidth);
```
Comment (carolynrehm) [inline src/app/components/reports-time-graph/reports-time-graph.component.ts:65]: Why was this default removed?

Comment (watchdogoblivion) [reply src/app/components/reports-time-graph/reports-time-graph.component.ts:65]: I was testing something and forgot to put it back. I will handle that.

File: src/app/services/reports.service.ts (hunk 22 above)
Comment (watchdogoblivion) [reply src/app/services/reports.service.ts:22]: Yes, it needed to be accessed for a few tests.

File: src/app/services/sort-search.service.ts (hunk 25 above)
Comment (watchdogoblivion) [reply src/app/services/sort-search.service.ts]: It is not necessary.

File: src/app/services/pages.service.ts (hunk 20 above)
Comment (watchdogoblivion) [reply src/app/services/pages.service.ts]: I placed a comment over a test related to this change explaining this. If you want me to reset it and remove the test, I will do so.

File: src/app/components/reports-page/reports-page.component.ts (hunk 8 above)
Comment (watchdogoblivion) [reply src/app/components/reports-page/reports-page.component.ts:2]: Understood. I will try not to remove them.

File: src/app/components/reports-time-graph/reports-time-graph.component.ts (hunk 27 above)
Comment (watchdogoblivion) [reply src/app/components/reports-time-graph/reports-time-graph.component.ts:65]: I was going through the files again, and I just remembered why I removed the default. The documentation for window states that window.innerWidth does not have a default value. I assume this means that window.innerWidth can be null or some other falsy. However, in the case that it is falsy, in this context, Math.max(any number, falsy) will treat the falsy as 0. So there is no point in using the guard. Unless I misread this.

File: src/app/services/pages.service.ts (hunk 20 above)
Comment (watchdogoblivion) [reply src/app/services/pages.service.ts]: It appears the service only serves the purpose of hosting a centralized means of refreshing the page with the common window.location.reload. Since the window cannot be spied on and there is no other logic or functionality in the service, I reset everything per your request and simply removed the service from code coverage. This comment also applies to the previous comment from you and Carolyn regarding asana, associated with a comment left in the complement spec file.

**Comment by carolynrehm** [comment]
@watchdogoblivion please remove src/app/services/sort-search.service.ts and fixing merge conflicts and we'll work on getting this merged in

**Review by IkenoXamos — DISMISSED** [review]

**Review by carolynrehm — APPROVED** [review]
